### PR TITLE
KNN support for spoints

### PR DESCRIPTION
**Pull request by vitcpp**
This is the original patch for review and improvement.

### CODE REVIEW

File: doc/indices.sgm [hunk 1]
```diff
@@ -68,6 +68,11 @@
             </para>
           </listitem>
         </itemizedlist>
+        <para>
+          GiST index can be used also for fast finding points closest to the given one
```
Comment (esabol) [inline doc/indices.sgm:72]: Change "fast finding points" to "quickly finding the points".

File: doc/indices.sgm [hunk 2]
```diff
@@ -68,6 +68,11 @@
             </para>
           </listitem>
         </itemizedlist>
+        <para>
+          GiST index can be used also for fast finding points closest to the given one 
+          when ordering by an expression with the <literal>&lt;-&gt;</literal> operator
+          is used, as shown in an example below.
```
Comment (esabol) [inline doc/indices.sgm:74]: Remove "is used" and move the comma to be at the end of "operator" on the previous line.

File: doc/indices.sgm [hunk 3]
```diff
@@ -100,7 +112,6 @@
 <![CDATA[CREATE INDEX test_pos_idx USING BRIN ON test (pos) WITH (pages_per_range = 16);]]>
           </programlisting>
         </example>
-
```
Comment (esabol) [inline doc/indices.sgm:103]: Based on the other `</sect1>` tags I saw, I don't this blank line before `</sect1>` should have been removed.

File: pgs_gist.sql.in [hunk 4]
```diff
@@ -98,12 +98,15 @@ CREATE FUNCTION g_spoint_compress(internal)
    AS 'MODULE_PATHNAME', 'g_spoint_compress'
    LANGUAGE 'c';
 
-
 CREATE FUNCTION g_spoint_consistent(internal, internal, int4, oid, internal)
    RETURNS internal
    AS 'MODULE_PATHNAME', 'g_spoint_consistent'
    LANGUAGE 'c';
 
+CREATE FUNCTION g_spoint_distance(internal, spoint, smallint, oid, internal)
```
Comment (esabol) [inline pgs_gist.sql.in:106]: Create an upgrade script and add this function to it.

File: pgs_gist.sql.in [hunk 5]
```diff
@@ -114,6 +117,7 @@ CREATE OPERATOR CLASS spoint
    OPERATOR  14 @ (spoint, spoly),
    OPERATOR  15 @ (spoint, sellipse),
    OPERATOR  16 @ (spoint, sbox),
+   OPERATOR	 17	<-> (spoint, spoint) FOR ORDER BY float_ops,
```
Comment (esabol) [inline pgs_gist.sql.in:120]: Whitespace doesn't match the other `OPERATOR` lines surrounding this addition. The `17` should line up with the `16` on the previous line for starters.

File: src/gist.c [hunk 6]
```diff
@@ -681,6 +682,10 @@ g_spoint3_consistent(PG_FUNCTION_ARGS)
 	PG_RETURN_BOOL(false);
 }
 
+static double distance_vector_point_3d (Vector3D* v, double x, double y, double z) {
```
Comment (esabol) [inline src/gist.c:685]: Change `Vector3D* v` to `Vector3D *v` in order to match the coding style of pgsphere.

File: src/gist.c [hunk 7]
```diff
@@ -681,6 +682,10 @@ g_spoint3_consistent(PG_FUNCTION_ARGS)
 	PG_RETURN_BOOL(false);
 }
 
+static double distance_vector_point_3d (Vector3D* v, double x, double y, double z) { 
+	return acos ( (v->x * x + v->y * y + v->z * z) / sqrt( x*x + y*y + z*z ) ); //  as v has length=1 by design
```
Comment (esabol) [inline src/gist.c:686]: Remove the space after `acos`.

File: src/gist.c [hunk 8]
```diff
@@ -1672,6 +1677,127 @@ fallbackSplit(Box3D *boxes, OffsetNumber maxoff, GIST_SPLITVEC *v)
...
+			// Prepare data for calculation
+			spoint_vector3d(&v_point, point);
+			v_low.x  = (double)box->low.coord[0]  / MAXCVALUE;
+			v_low.y  = (double)box->low.coord[1]  / MAXCVALUE;
+			v_low.z  = (double)box->low.coord[2]  / MAXCVALUE;
+			v_high.x = (double)box->high.coord[0]  / MAXCVALUE;
+			v_high.y = (double)box->high.coord[1]  / MAXCVALUE;
+			v_high.z = (double)box->high.coord[2]  / MAXCVALUE;
+			//  a box splits space into 27 subspaces (6+12+8+1) with different distance calculation
+			if(v_point.x < v_low.x) {
```
Comment (esabol) [inline src/gist.c:1703]: Change `if(` to `if (` throughout this function in order to match the coding style of pgsphere.

File: src/gist.c [hunk 9]
```diff
@@ -1672,6 +1677,127 @@ fallbackSplit(Box3D *boxes, OffsetNumber maxoff, GIST_SPLITVEC *v)
...
+			v_low.y  = (double)box->low.coord[1]  / MAXCVALUE;
+			v_low.z  = (double)box->low.coord[2]  / MAXCVALUE;
+			v_high.x = (double)box->high.coord[0]  / MAXCVALUE;
+			v_high.y = (double)box->high.coord[1]  / MAXCVALUE;
+			v_high.z = (double)box->high.coord[2]  / MAXCVALUE;
+			//  a box splits space into 27 subspaces (6+12+8+1) with different distance calculation
+			if(v_point.x < v_low.x) { 
+				if(v_point.y < v_low.y) { 
+					if(v_point.z < v_low.z) { 
+						retval = distance_vector_point_3d (&v_point, v_low.x, v_low.y, v_low.z); //point2point distance
```
Comment (esabol) [inline src/gist.c:1706]: Change `distance_vector_point_3d (` to `distance_vector_point_3d(` throughout this function in order to match the coding style of pgsphere.

File: src/gist.c [hunk 10]
```diff
@@ -1672,6 +1677,127 @@ fallbackSplit(Box3D *boxes, OffsetNumber maxoff, GIST_SPLITVEC *v)
 	v->spl_ldatum_exists = v->spl_rdatum_exists = false;
 }
 
+
```
Comment (esabol) [inline src/gist.c:1680]: An extra blank line doesn't need to be added here.

File: src/gist.c [hunk 11]
```diff
@@ -1672,6 +1677,127 @@ fallbackSplit(Box3D *boxes, OffsetNumber maxoff, GIST_SPLITVEC *v)
...
+		default:
+			elog(ERROR, "unrecognized cube strategy number: %d", strategy);
+			retval = 0;		/* keep compiler quiet */
+			break;
+	}
+	PG_RETURN_FLOAT8(retval);
+}
+
+
```
Comment (esabol) [inline src/gist.c:1799]: Two too many blank lines added here, I think.

**Review by esabol — CHANGES_REQUESTED** [review]

**Comment by vitcpp** [comment]
I used pgindent to reformat new functions in gist.c.

File: doc/indices.sgm [hunk 12]
```diff
@@ -68,6 +68,11 @@
             </para>
           </listitem>
         </itemizedlist>
+        <para>
+          GiST index can be also used for quickly finding the points closest to the given one
```
Comment (esabol) [inline doc/indices.sgm:72]: Change `GiST index` to `A GiST index`. Sorry should have caught that in my previous review.

File: doc/indices.sgm [hunk 13]
```diff
@@ -82,6 +87,13 @@
 <![CDATA[CREATE INDEX test_pos_idx ON test USING GIST (pos);]]>
 <![CDATA[VACUUM ANALYZE test;]]>
           </programlisting>
+          <para>
+           To find points closest to a given spherical position, use the <literal>&lt;-&gt;</literal> operator:
```
Comment (esabol) [inline doc/indices.sgm:91]: Change `To find points` to `To find the points`.

**Review by esabol — CHANGES_REQUESTED** [review]
A couple more very minor changes to the documentation, please.

Still needs an upgrade script. And maybe a version bump?

File: upgrade_scripts/pg_sphere--1.4.2--1.5.0.sql.in [hunk 14]
```diff
@@ -0,0 +1,16 @@
+-- Upgrade: 1.4.2 -> 1.5.0
+
+CREATE FUNCTION g_spoint_distance(internal, spoint, smallint, oid, internal)
```
Comment (esabol) [inline upgrade_scripts/pg_sphere--1.4.2--1.5.0.sql.in:3]: Use `CREATE OR REPLACE` in the upgrade script? See issue #111.

**Review by esabol — CHANGES_REQUESTED** [review]

**Review by esabol — APPROVED** [review]
Looks good to me!

**Comment by vitcpp** [comment]
Squashed latest 3 commits. The two commits are kept - the original commit and the commit with code adjustment.

**Pull request merged**